src/main/java/org/commonjava/indy/service/archive/controller/ArchiveController.java

```diff
@@ -0,0 +1,319 @@
+    {
+        BasicCookieStore cookieStore = new BasicCookieStore();
+        ExecutorCompletionService<Boolean> executor = new ExecutorCompletionService<>( executorService );
+
+        String contentBuildDir = String.format( "%s/%s", contentDir, content.getBuildConfigId() );
+        File dir = new File( contentBuildDir );
+        dir.delete();
+
+        fileTrackedContent( contentBuildDir, content );
+
+        for ( String path : downloadPaths.keySet() )
+        {
+            String filePath = downloadPaths.get( path );
+            executor.submit( download( contentBuildDir, path, filePath, cookieStore ) );
+        }
+        int success = 0;
+        int failed = 0;
+        for ( int i = 0; i < downloadPaths.size(); i++ )
+        {
+            if ( executor.take().get() )
+            {
+                success++;
+            }
+            else
+            {
+                failed++;
+            }
+        }
+        logger.info( "Artifacts download completed, success:{}, failed:{}", success, failed );
+    }
+
+    public Optional<File> generateArchive( final HistoricalContentDTO content )
+            throws IOException
+    {
+        String contentBuildDir = String.format( "%s/%s", contentDir, content.getBuildConfigId() );
+        File dir = new File( contentBuildDir );
+        if ( !dir.exists() )
+        {
+            return Optional.empty();
+        }
+
+        final File part = new File( archiveDir, content.getBuildConfigId() + PART_ARCHIVE_SUFFIX );
+        part.getParentFile().mkdirs();
+
+        logger.info( "Writing archive to: '{}'", part.getAbsolutePath() );
+        ZipOutputStream zip = new ZipOutputStream( new FileOutputStream( part ) );
+        List<File> artifacts = walkedAllFiles( contentBuildDir );
+
+        byte[] buffer = new byte[1024];
+        for ( File artifact : artifacts )
+        {
+            logger.trace( "Adding {} to archive {} in folder {}", artifact.getName(), part.getName(), archiveDir );
+            FileInputStream fis = new FileInputStream( artifact );
+            String entryPath = artifact.getPath().split( contentBuildDir )[1];
+
+            zip.putNextEntry( new ZipEntry( entryPath ) );
+
+            int length;
+            while ( ( length = fis.read(buffer ) ) > 0 )
+            {
+                zip.write( buffer, 0, length );
+            }
+            zip.closeEntry();
+            fis.close();
+        }
+        zip.close();
+
+        //clean obsolete build contents
+        for ( File artifact : artifacts )
+        {
+            Files.delete( artifact.toPath() );
+        }
+        dir.delete();
+        return Optional.of( part );
+    }
+
+    public void renderArchive( File part, final String buildConfigId )
+    {
+        if ( !part.exists() )
+        {
+            return;
+        }
+        final File target = new File( archiveDir, buildConfigId + ARCHIVE_SUFFIX );
+        target.delete();
```

Comment:
I'm thinking a problem here for a concurrent problem. If a new archive generation completed during some downloading of the old one still happening, this will cause problem.  I think we should consider the result of this delete action, and only when deletion succeeded then do the rename action.
And should we add some concurrent tests here to see what will happen for this case?

Reply:
maybe we should add independent dir for different buildconfig, it's more viable for the codes here, since generating archive not only will include other concurrent archive file, the file delete part will also be fatal.
the best is to use some file lock, I can investigate and do this improvement tho.

Reply:
@ligangty It might be a boundary case? I'm not sure if PNC will build from the old archive in progress meanwhile a new archive is generated on road after promotion.
Could we confirm with PNC then create a story to resolve such case?